src/backends/plonky2/primitives/merkletree/circuit.rs

```diff
@@ -385,6 +391,234 @@ fn kv_hash_target(
+        proof.enabled.target,
+    );
+
+    // 4) assert proof_non_existence.root==old_root, and that it uses new_key &
+    // new_value
+    for j in 0..HASH_SIZE {
+        // 4.1) assert that proof.proof_non_existence.root == proof.old_root
+        builder.conditional_assert_eq(
+            proof.enabled.target,
+            proof.proof_non_existence.root.elements[j],
+            proof.old_root.elements[j],
+        );
+        // 4.2) assert that the non-existence proof uses the new_key & new_value
+        builder.conditional_assert_eq(
+            proof.enabled.target,
+            proof.proof_non_existence.key.elements[j],
+            proof.new_key.elements[j],
+        );
+        builder.conditional_assert_eq(
+            proof.enabled.target,
+            proof.proof_non_existence.value.elements[j],
+            proof.new_value.elements[j],
+        );
+    }
+
+    // prepare value for check 5.2)
+    let old_leaf_hash = kv_hash_target(
+        builder,
+        &proof.proof_non_existence.other_key,
+        &proof.proof_non_existence.other_value,
+    );
+    // prepare values for check 5.3)
+    let old_leaf_path = keypath_target(
+        proof.max_depth,
+        builder,
+        &proof.proof_non_existence.other_key,
+    );
+
+    // 5) check that old_siblings & new_siblings match as expected. Let
+    //    d=divergence_level, assert that:
+    // 5.1) old_siblings[i] == new_siblings[i] ∀ i \ {d}
+    // 5.2) at i==d, if old_siblings[i] != new_siblings[i]: old_siblings[i] ==
+    //   EMPTY_HASH new_siblings[i] == old_leaf_hash
+    // 5.3) assert that if old_key!=empty, both old_leaf_path&new_leaf_path
+    //   should diverge at the inputted divergence level
+    let old_siblings = proof.proof_non_existence.siblings.clone();
+    let new_siblings = proof.new_siblings.clone();
+    for i in 0..proof.max_depth {
+        let i_targ = builder.constant(F::from_canonical_u64(i as u64));
+        let is_divergence_level = builder.is_equal(i_targ, proof.divergence_level);
+
+        // 5.1) for all i except for i==divergence_level, assert that the
+        //   siblings are the same
+        let old_sibling_i: Vec<Target> = (0..HASH_SIZE)
+            .map(|j| builder.select(is_divergence_level, zero, old_siblings[i].elements[j]))
+            .collect();
+        let new_sibling_i: Vec<Target> = (0..HASH_SIZE)
+            .map(|j| builder.select(is_divergence_level, zero, new_siblings[i].elements[j]))
+            .collect();
+        for j in 0..HASH_SIZE {
+            builder.conditional_assert_eq(proof.enabled.target, old_sibling_i[j], new_sibling_i[j]);
+        }
+
+        // 5.2) when i==d && if old_siblings[i] != new_siblings[i], check that:
+        //   old_siblings[i] == EMPTY_HASH && new_siblings[i] == old_leaf_hash
+
+        // in_case_5_2=true if: i==d (= is_divergence_level) && old_siblings[i]!=new_siblings[i]
+        let old_is_eq_new = zip_eq(old_siblings[i].elements, new_siblings[i].elements).fold(
```

Comment:
you could use `reduce` (https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.reduce) instead of `fold` here because the len of siblings is always > 0.